File: src/modules/coupon/coupon.service.ts
import { Types } from 'mongoose';
import { AppError } from '../../shared/errors/AppError.js';
import { recordAudit } from '../../shared/utils/audit.js';
import {
  buildPaginationMeta,
  toPageOptions,
  type PaginationMeta,
} from '../../shared/utils/pagination.js';
import { Coupon, type CouponType, type CouponScope } from './coupon.model.js';
import { CouponUsage } from './coupon-usage.model.js';

export interface CouponPublic {
  id: string;
  code: string;
  type: CouponType;
  value: number;
  scope: CouponScope;
  productIds: string[];
  categoryIds: string[];
  minOrderValueCents: number;
  maxDiscountCents: number | null;
  maxUses: number | null;
  perUserLimit: number;
  usedCount: number;
  validFrom: string;
  validUntil: string;
  isActive: boolean;
}

function toCouponPublic(coupon: {
  _id: unknown;
  code: string;
  type: CouponType;
  value: number;
  scope: CouponScope;
  productIds: Array<{ toString(): string }>;
  categoryIds: Array<{ toString(): string }>;
  minOrderValueCents: number;
  maxDiscountCents: number | null;
  maxUses: number | null;
  perUserLimit: number;
  usedCount: number;
  validFrom: Date;
  validUntil: Date;
  isActive: boolean;
}): CouponPublic {
  return {
    id: (coupon._id as { toString(): string }).toString(),
    code: coupon.code,
    type: coupon.type,
    value: coupon.value,
    scope: coupon.scope,
    productIds: coupon.productIds.map((p) => p.toString()),
    categoryIds: coupon.categoryIds.map((c) => c.toString()),
    minOrderValueCents: coupon.minOrderValueCents,
    maxDiscountCents: coupon.maxDiscountCents,
    maxUses: coupon.maxUses,
    perUserLimit: coupon.perUserLimit,
    usedCount: coupon.usedCount,
    validFrom: coupon.validFrom.toISOString(),
    validUntil: coupon.validUntil.toISOString(),
    isActive: coupon.isActive,
  };
}

export async function createCoupon(
  input: {
    code: string;
    type: CouponType;
    value: number;
    scope?: CouponScope;
    productIds?: string[];
    categoryIds?: string[];
    minOrderValueCents?: number;
    maxDiscountCents?: number | null;
    maxUses?: number | null;
    perUserLimit?: number;
    validFrom?: Date | string;
    validUntil?: Date | string;
    isActive?: boolean;
  },
  actorId: string,
): Promise<CouponPublic> {
  const code = input.code.trim().toUpperCase();
  const coupon = await Coupon.create({
    code,
    type: input.type,
    value: input.value,
    scope: input.scope ?? 'all',
    productIds: input.productIds ?? [],
    categoryIds: input.categoryIds ?? [],
    minOrderValueCents: input.minOrderValueCents ?? 0,
    maxDiscountCents: input.maxDiscountCents ?? null,
    maxUses: input.maxUses ?? null,
    perUserLimit: input.perUserLimit ?? 1,
    usedCount: 0,
    validFrom: input.validFrom ? new Date(input.validFrom) : new Date(0),
    validUntil: input.validUntil ? new Date(input.validUntil) : new Date('2099-12-31T23:59:59Z'),
    isActive: input.isActive ?? true,
  });
  recordAudit('coupon.created', { couponId: coupon._id.toString(), code }, { actorId });
  return toCouponPublic(coupon);
}

export async function updateCoupon(
  couponId: string,
  input: Partial<Parameters<typeof createCoupon>[0]>,
  actorId: string,
): Promise<CouponPublic> {
  const coupon = await Coupon.findById(couponId);
  if (!coupon) throw AppError.notFound('Coupon not found');

  if (input.code !== undefined) coupon.code = input.code.trim().toUpperCase();
  if (input.type !== undefined) coupon.type = input.type;
  if (input.value !== undefined) coupon.value = input.value;
  if (input.scope !== undefined) coupon.scope = input.scope;
  if (input.productIds !== undefined)
    coupon.productIds = input.productIds.map((p) => new Types.ObjectId(p));
  if (input.categoryIds !== undefined)
    coupon.categoryIds = input.categoryIds.map((c) => new Types.ObjectId(c));
  if (input.minOrderValueCents !== undefined) coupon.minOrderValueCents = input.minOrderValueCents;
  if (input.maxDiscountCents !== undefined) coupon.maxDiscountCents = input.maxDiscountCents;
  if (input.maxUses !== undefined) coupon.maxUses = input.maxUses;
  if (input.perUserLimit !== undefined) coupon.perUserLimit = input.perUserLimit;
  if (input.validFrom !== undefined) coupon.validFrom = new Date(input.validFrom as Date | string);
  if (input.validUntil !== undefined)
    coupon.validUntil = new Date(input.validUntil as Date | string);
  if (input.isActive !== undefined) coupon.isActive = input.isActive;

  await coupon.save();
  recordAudit('coupon.updated', { couponId }, { actorId });
  return toCouponPublic(coupon);
}

export async function deleteCoupon(couponId: string, actorId: string): Promise<void> {
  const result = await Coupon.deleteOne({ _id: couponId });
  if (result.deletedCount === 0) throw AppError.notFound('Coupon not found');
  recordAudit('coupon.deleted', { couponId }, { actorId });
}

export async function listCoupons(query: {
  page?: number;
  limit?: number;
  isActive?: 'true' | 'false';
}): Promise<{ coupons: CouponPublic[]; meta: PaginationMeta }> {
  const filter: Record<string, unknown> = {};
  if (query.isActive) filter.isActive = query.isActive === 'true';
  const options = toPageOptions(query);
  const [total, coupons] = await Promise.all([
    Coupon.countDocuments(filter),
    Coupon.find(filter).sort({ createdAt: -1 }).skip(options.skip).limit(options.limit).lean(),
  ]);
  return {
    coupons: coupons.map((c) =>
      toCouponPublic(c as unknown as Parameters<typeof toCouponPublic>[0]),
    ),
    meta: buildPaginationMeta(total, options),
  };
}

// ---------------------------------------------------------------------------
// Discount evaluation during checkout
// ---------------------------------------------------------------------------

export interface DiscountCartItem {
  productId: string;
  category: string | null;
  unitPriceCents: number;
  quantity: number;
}

export interface DiscountResult {
  code: string;
  couponId: string;
  discountCents: number;
}

// Pure computation, no DB writes: the order module re-checks limits inside its
// transaction before committing the coupon usage.
export async function computeDiscount(
  couponCode: string,
  userId: string,
  items: DiscountCartItem[],
): Promise<DiscountResult> {
  const coupon = await Coupon.findOne({ code: couponCode.trim().toUpperCase() });
  if (!coupon) throw AppError.badRequest('Invalid coupon code', { errorCode: 'INVALID_COUPON' });
  if (!coupon.isActive)
    throw AppError.badRequest('This coupon is no longer active', { errorCode: 'COUPON_INACTIVE' });

  const now = new Date();
  if (coupon.validFrom > now)
    throw AppError.badRequest('This coupon is not yet valid', { errorCode: 'COUPON_NOT_STARTED' });
  if (coupon.validUntil < now)
    throw AppError.badRequest('This coupon has expired', { errorCode: 'COUPON_EXPIRED' });

  const subtotal = items.reduce((sum, i) => sum + i.unitPriceCents * i.quantity, 0);
  if (subtotal < coupon.minOrderValueCents) {
    throw AppError.badRequest(
      `Minimum order value for this coupon is ${(coupon.minOrderValueCents / 100).toFixed(2)}`,
      { errorCode: 'COUPON_MIN_ORDER_NOT_MET' },
    );
  }

  if (coupon.maxUses !== null && coupon.usedCount >= coupon.maxUses) {
    throw AppError.badRequest('This coupon has reached its usage limit', {
      errorCode: 'COUPON_EXHAUSTED',
    });
  }

  const userUsage = await CouponUsage.countDocuments({ couponId: coupon._id, userId });
  if (userUsage >= coupon.perUserLimit) {
    throw AppError.badRequest('You have already used this coupon', {
      errorCode: 'COUPON_USER_LIMIT_REACHED',
    });
  }

  // Eligible subtotal depends on the coupon scope. For scoped coupons, the
  // discount only applies to matching items; if nothing matches, the coupon
  // simply cannot be used with this cart.
  let eligibleSubtotalCents = 0;
  for (const item of items) {
    const lineTotal = item.unitPriceCents * item.quantity;
    if (coupon.scope === 'all') {
      eligibleSubtotalCents += lineTotal;
    } else if (coupon.scope === 'category' && item.category) {
      if (coupon.categoryIds.some((id) => id.toString() === item.category))
        eligibleSubtotalCents += lineTotal;
    } else if (coupon.scope === 'product') {
      if (coupon.productIds.some((id) => id.toString() === item.productId))
        eligibleSubtotalCents += lineTotal;
    }
  }

  if (eligibleSubtotalCents <= 0) {
    throw AppError.badRequest('This coupon does not apply to any item in your cart', {
      errorCode: 'COUPON_NOT_APPLICABLE',
    });
  }

  let discountCents =
    coupon.type === 'percentage'
      ? Math.floor((eligibleSubtotalCents * coupon.value) / 100)
      : coupon.value;

  if (coupon.maxDiscountCents !== null)
    discountCents = Math.min(discountCents, coupon.maxDiscountCents);
  discountCents = Math.min(discountCents, eligibleSubtotalCents);

  return { code: coupon.code, couponId: coupon._id.toString(), discountCents };
}

File: java/src/main/java/com/ecommerce/backend/modules/coupon/CouponService.java
package com.ecommerce.backend.modules.coupon;

import com.ecommerce.backend.common.api.PageMeta;
import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.modules.audit.AuditService;
import com.ecommerce.backend.modules.coupon.CouponDtos.CouponInput;
import com.ecommerce.backend.modules.coupon.CouponDtos.CouponUpdate;
import com.ecommerce.backend.modules.coupon.CouponDtos.ValidateCouponRequest.ValidateItem;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import org.bson.types.ObjectId;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

/** Coupon business logic (mirrors {@code coupon.service.ts}). */
@Service
public class CouponService {

    private static final Instant DEFAULT_VALID_FROM = Instant.ofEpochMilli(0);
    private static final Instant DEFAULT_VALID_UNTIL = Instant.parse("2099-12-31T23:59:59Z");

    private final CouponRepository couponRepository;
    private final CouponUsageRepository couponUsageRepository;
    private final AuditService auditService;

    public CouponService(
            CouponRepository couponRepository,
            CouponUsageRepository couponUsageRepository,
            AuditService auditService) {
        this.couponRepository = couponRepository;
        this.couponUsageRepository = couponUsageRepository;
        this.auditService = auditService;
    }

    public record CouponPublic(
            String id,
            String code,
            String type,
            long value,
            String scope,
            List<String> productIds,
            List<String> categoryIds,
            long minOrderValueCents,
            Long maxDiscountCents,
            Long maxUses,
            int perUserLimit,
            long usedCount,
            String validFrom,
            String validUntil,
            boolean isActive) {}

    // -------------------------------------------------------------- Management

    public CouponPublic createCoupon(CouponInput input, String actorId) {
        Coupon coupon = new Coupon();
        coupon.code = input.code().trim().toUpperCase(Locale.ROOT);
        coupon.type = input.type();
        coupon.value = input.value();
        coupon.scope = input.scope() == null ? Coupon.SCOPE_ALL : input.scope();
        coupon.productIds = stringsOrEmpty(input.productIds());
        coupon.categoryIds = stringsOrEmpty(input.categoryIds());
        coupon.minOrderValueCents = input.minOrderValueCents() == null ? 0 : input.minOrderValueCents();
        coupon.maxDiscountCents = input.maxDiscountCents();
        coupon.maxUses = input.maxUses();
        coupon.perUserLimit = input.perUserLimit() == null ? 1 : input.perUserLimit();
        coupon.usedCount = 0;
        coupon.validFrom = input.validFrom() == null || input.validFrom().isBlank()
                ? DEFAULT_VALID_FROM
                : Instant.parse(input.validFrom());
        coupon.validUntil = input.validUntil() == null || input.validUntil().isBlank()
                ? DEFAULT_VALID_UNTIL
                : Instant.parse(input.validUntil());
        coupon.isActive = input.isActive() == null ? true : input.isActive();
        coupon.createdAt = Instant.now();
        coupon.updatedAt = coupon.createdAt;

        Coupon saved = couponRepository.save(coupon);
        auditService.log("coupon.created", actorId, null,
                java.util.Map.of("couponId", saved.id, "code", saved.code));
        return toPublic(saved);
    }

    public CouponPublic updateCoupon(String couponId, CouponUpdate input, String actorId) {
        Coupon coupon = requireCoupon(couponId);
        if (input.code() != null) {
            coupon.code = input.code().trim().toUpperCase(Locale.ROOT);
        }
        if (input.type() != null) {
            coupon.type = input.type();
        }
        if (input.value() != null) {
            coupon.value = input.value();
        }
        if (input.scope() != null) {
            coupon.scope = input.scope();
        }
        if (input.productIds() != null) {
            coupon.productIds = stringsOrEmpty(input.productIds());
        }
        if (input.categoryIds() != null) {
            coupon.categoryIds = stringsOrEmpty(input.categoryIds());
        }
        if (input.minOrderValueCents() != null) {
            coupon.minOrderValueCents = input.minOrderValueCents();
        }
        if (input.maxDiscountCents() != null) {
            coupon.maxDiscountCents = input.maxDiscountCents();
        }
        if (input.maxUses() != null) {
            coupon.maxUses = input.maxUses();
        }
        if (input.perUserLimit() != null) {
            coupon.perUserLimit = input.perUserLimit();
        }
        if (input.validFrom() != null) {
            coupon.validFrom = Instant.parse(input.validFrom());
        }
        if (input.validUntil() != null) {
            coupon.validUntil = Instant.parse(input.validUntil());
        }
        if (input.isActive() != null) {
            coupon.isActive = input.isActive();
        }
        coupon.updatedAt = Instant.now();
        Coupon saved = couponRepository.save(coupon);
        auditService.log("coupon.updated", actorId, null, java.util.Map.of("couponId", saved.id));
        return toPublic(saved);
    }

    public void deleteCoupon(String couponId, String actorId) {
        if (!couponRepository.existsById(couponId)) {
            throw ApiException.notFound("Coupon not found");
        }
        couponRepository.deleteById(couponId);
        auditService.log("coupon.deleted", actorId, null, java.util.Map.of("couponId", couponId));
    }

    public record CouponListResult(List<CouponPublic> coupons, long total) {}

    public CouponListResult listCoupons(int page, int limit, Boolean isActive) {
        long total = couponRepository.count();
        List<Coupon> coupons = couponRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"))
                .stream()
                .filter(c -> isActive == null || c.isActive == isActive)
                .skip((long) (page - 1) * limit)
                .limit(limit)
                .toList();
        return new CouponListResult(coupons.stream().map(this::toPublic).toList(), total);
    }

    // -------------------------------------------------------------- Discount

    public record DiscountCartItem(String productId, String category, long unitPriceCents, int quantity) {}

    public record DiscountResult(String code, String couponId, long discountCents) {}

    /** Pure computation, no DB writes; the order module re-checks limits before committing usage. */
    public DiscountResult computeDiscount(String couponCode, String userId, List<DiscountCartItem> items) {
        Coupon coupon = couponRepository
                .findByCode(couponCode.trim().toUpperCase(Locale.ROOT))
                .orElseThrow(() -> ApiException.badRequest("Invalid coupon code", "INVALID_COUPON"));
        if (!coupon.isActive) {
            throw ApiException.badRequest("This coupon is no longer active", "COUPON_INACTIVE");
        }

        Instant now = Instant.now();
        if (coupon.validFrom != null && coupon.validFrom.isAfter(now)) {
            throw ApiException.badRequest("This coupon is not yet valid", "COUPON_NOT_STARTED");
        }
        if (coupon.validUntil != null && coupon.validUntil.isBefore(now)) {
            throw ApiException.badRequest("This coupon has expired", "COUPON_EXPIRED");
        }

        long subtotal = items.stream().mapToLong(i -> i.unitPriceCents() * i.quantity()).sum();
        if (subtotal < coupon.minOrderValueCents) {
            throw ApiException.badRequest(
                    "Minimum order value for this coupon is "
                            + String.format("%.2f", coupon.minOrderValueCents / 100.0),
                    "COUPON_MIN_ORDER_NOT_MET");
        }

        if (coupon.maxUses != null && coupon.usedCount >= coupon.maxUses) {
            throw ApiException.badRequest(
                    "This coupon has reached its usage limit", "COUPON_EXHAUSTED");
        }

        long userUsage = couponUsageRepository.countByCouponIdAndUserId(coupon.id, userId);
        if (userUsage >= coupon.perUserLimit) {
            throw ApiException.badRequest(
                    "You have already used this coupon", "COUPON_USER_LIMIT_REACHED");
        }

        long eligibleSubtotalCents = 0;
        for (DiscountCartItem item : items) {
            long lineTotal = item.unitPriceCents() * item.quantity();
            if (Coupon.SCOPE_ALL.equals(coupon.scope)) {
                eligibleSubtotalCents += lineTotal;
            } else if (Coupon.SCOPE_CATEGORY.equals(coupon.scope) && item.category() != null) {
                if (coupon.categoryIds.contains(item.category())) {
                    eligibleSubtotalCents += lineTotal;
                }
            } else if (Coupon.SCOPE_PRODUCT.equals(coupon.scope)) {
                if (coupon.productIds.contains(item.productId())) {
                    eligibleSubtotalCents += lineTotal;
                }
            }
        }

        if (eligibleSubtotalCents <= 0) {
            throw ApiException.badRequest(
                    "This coupon does not apply to any item in your cart", "COUPON_NOT_APPLICABLE");
        }

        long discountCents = Coupon.TYPE_PERCENTAGE.equals(coupon.type)
                ? Math.floorDiv(eligibleSubtotalCents * coupon.value, 100)
                : coupon.value;
        if (coupon.maxDiscountCents != null) {
            discountCents = Math.min(discountCents, coupon.maxDiscountCents);
        }
        discountCents = Math.min(discountCents, eligibleSubtotalCents);

        return new DiscountResult(coupon.code, coupon.id, discountCents);
    }

    // -------------------------------------------------------------- Internals

    public Coupon requireCoupon(String couponId) {
        if (!ObjectId.isValid(couponId)) {
            throw ApiException.badRequest("Invalid identifier format", "BAD_REQUEST");
        }
        return couponRepository
                .findById(couponId)
                .orElseThrow(() -> ApiException.notFound("Coupon not found"));
    }

    private CouponPublic toPublic(Coupon c) {
        return new CouponPublic(
                c.id,
                c.code,
                c.type,
                c.value,
                c.scope,
                c.productIds == null ? List.of() : c.productIds,
                c.categoryIds == null ? List.of() : c.categoryIds,
                c.minOrderValueCents,
                c.maxDiscountCents,
                c.maxUses,
                c.perUserLimit,
                c.usedCount,
                iso(c.validFrom),
                iso(c.validUntil),
                c.isActive);
    }

    private static String iso(Instant value) {
        return value == null ? null : value.toString();
    }

    private static List<String> stringsOrEmpty(List<String> values) {
        return values == null ? new ArrayList<>() : values;
    }

    public PageMeta meta(int page, int limit, long total) {
        return PageMeta.of(page, limit, total);
    }

    public record ValidateCartItem(String productId, String category) {}
}
